### ISSUE TITLE
Scatter in ASR training causes error: Op mismatch in shape of indices and update value at dim#0 

### ISSUE THREAD

**Issue by phoenix-meadowlark** [iree-org/iree#11877]
### What happened?

Attempting to compile the gradient computation for an ASR model causes the following error:

```mlir
/tmp/asr_toy_grad.mlir:16:10: error: 'iree_linalg_ext.scatter' op mismatch in shape of indices and update value at dim#0
    %9 = "stablehlo.scatter"(%8, %1, %6) ({
         ^
```

### Steps to reproduce your issue

A toy reproducer can be generated from the following function:

```py
import jax
import jax.numpy as jnp

def asr_toy(w_1, w_2, x):
  y = jnp.einsum('abc, cd -> abd', x, w_1)
  y = y[:, 0]
  z = jnp.einsum('ab, bc -> ac', y, w_2)
  return jnp.sum(z)

w_1 = jnp.ones((4, 5))
w_2 = jnp.ones((5, 6))
x = jnp.ones((2, 3, 4))

ir_module = jax.jit(jax.grad(asr_toy), backend="iree").lower(w_1, w_2, x).compiler_ir()
with open("/tmp/asr_toy_grad.mlir", "wb") as f:
  ir_module.operation.print(file=f, binary=True)
```

MLIR:

```mlir
module @jit_asr_toy {
  func.func public @main(%arg0: tensor<5x6xf32> {mhlo.sharding = ""}, %arg1: tensor<2x3x4xf32> {mhlo.sharding = ""}) -> tensor<4x5xf32> {
    %0 = stablehlo.constant dense<0> : tensor<i32>
    %1 = stablehlo.broadcast_in_dim %0, dims = [] : (tensor<i32>) -> tensor<1xi32>
    %2 = stablehlo.constant dense<1.000000e+00> : tensor<f32>
    %3 = stablehlo.broadcast_in_dim %2, dims = [] : (tensor<f32>) -> tensor<2x6xf32>
    %4 = call @_einsum(%arg0, %3) : (tensor<5x6xf32>, tensor<2x6xf32>) -> tensor<2x5xf32>
    %5 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %6 = stablehlo.reduce(%4 init: %5) across dimensions = [] : (tensor<2x5xf32>, tensor<f32>) -> tensor<2x5xf32>
     reducer(%arg2: tensor<f32>, %arg3: tensor<f32>)  {
      %11 = stablehlo.add %arg2, %arg3 : tensor<f32>
      stablehlo.return %11 : tensor<f32>
    }
    %7 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %8 = stablehlo.broadcast_in_dim %7, dims = [] : (tensor<f32>) -> tensor<2x3x5xf32>
    %9 = "stablehlo.scatter"(%8, %1, %6) ({
    ^bb0(%arg2: tensor<f32>, %arg3: tensor<f32>):
      %11 = stablehlo.add %arg2, %arg3 : tensor<f32>
      stablehlo.return %11 : tensor<f32>
    }) {indices_are_sorted = true, scatter_dimension_numbers = #stablehlo.scatter<update_window_dims = [0, 1], inserted_window_dims = [1], scatter_dims_to_operand_dims = [1]>, unique_indices = true} : (tensor<2x3x5xf32>, tensor<1xi32>, tensor<2x5xf32>) -> tensor<2x3x5xf32>
    %10 = call @_einsum_0(%arg1, %9) : (tensor<2x3x4xf32>, tensor<2x3x5xf32>) -> tensor<4x5xf32>
    return %10 : tensor<4x5xf32>
  }
  func.func private @_einsum(%arg0: tensor<5x6xf32>, %arg1: tensor<2x6xf32>) -> tensor<2x5xf32> {
    %0 = "stablehlo.dot_general"(%arg1, %arg0) {dot_dimension_numbers = #stablehlo.dot<lhs_contracting_dimensions = [1], rhs_contracting_dimensions = [1]>, precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>]} : (tensor<2x6xf32>, tensor<5x6xf32>) -> tensor<2x5xf32>
    return %0 : tensor<2x5xf32>
  }
  func.func private @_einsum_0(%arg0: tensor<2x3x4xf32>, %arg1: tensor<2x3x5xf32>) -> tensor<4x5xf32> {
    %0 = "stablehlo.dot_general"(%arg1, %arg0) {dot_dimension_numbers = #stablehlo.dot<lhs_contracting_dimensions = [0, 1], rhs_contracting_dimensions = [0, 1]>, precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>]} : (tensor<2x3x5xf32>, tensor<2x3x4xf32>) -> tensor<5x4xf32>
    %1 = stablehlo.transpose %0, dims = [1, 0] : (tensor<5x4xf32>) -> tensor<4x5xf32>
    return %1 : tensor<4x5xf32>
  }
}
```

```shell
iree-compile \
  --iree-hal-target-backends=llvm-cpu \
  --iree-input-type=mhlo \
  /tmp/asr_toy_grad.mlir \
  -o /tmp/asr_toy_grad.vmfb
```

Full error:

```mlir
/tmp/asr_toy_grad.mlir:16:10: error: 'iree_linalg_ext.scatter' op mismatch in shape of indices and update value at dim#0
    %9 = "stablehlo.scatter"(%8, %1, %6) ({
         ^
/tmp/asr_toy_grad.mlir:16:10: note: see current operation:
%18 = "iree_linalg_ext.scatter"(%15, %5, %17) ({
^bb0(%arg3: f32, %arg4: f32):
  %19 = "arith.addf"(%arg4, %arg3) {fastmath = #arith.fastmath<none>} : (f32, f32) -> f32
  "iree_linalg_ext.yield"(%19) : (f32) -> ()
}) {dimension_map = array<i64: 1>, lowering_config = #iree_codegen.lowering_config<tile_sizes = [[0, 1, 1, 1]]>, operand_segment_sizes = array<i32: 2, 1>, unique_indices = true} : (tensor<?x1x1x1xf32>, tensor<1x1xi32>, tensor<1x1x1xf32>) -> tensor<1x1x1xf32>
/tmp/asr_toy_grad.mlir:16:10: error: failed to run translation of source executable to target executable for backend #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
    %9 = "stablehlo.scatter"(%8, %1, %6) ({
         ^
/tmp/asr_toy_grad.mlir:16:10: note: see current operation:
"hal.executable.variant"() ({
  "hal.executable.export"() ({
  ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index, %arg3: index, %arg4: index):
    "hal.return"(%arg4, %arg3, %arg2) : (index, index, index) -> ()
  }) {layout = #hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer>]>]>, ordinal = 0 : index, sym_name = "_main_dispatch_2_fill_2x3x5", translation_info = #iree_codegen.translation_info<CPUDefault>} : () -> ()
  "builtin.module"() ({
    "func.func"() ({
      %0 = "arith.constant"() {value = 0 : index} : () -> index
      %1 = "arith.constant"() {value = 1 : index} : () -> index
      %2 = "arith.constant"() {value = 2 : index} : () -> index
      %3 = "arith.constant"() {value = 5 : index} : () -> index
      %4 = "arith.constant"() {value = 128 : index} : () -> index
      %5 = "arith.constant"() {value = dense<0> : tensor<1x1xi32>} : () -> tensor<1x1xi32>
      %6 = "arith.constant"() {value = 0.000000e+00 : f32} : () -> f32
      %7 = "hal.interface.binding.subspan"(%4) {alignment = 64 : index, binding = 0 : index, descriptor_type = #hal.descriptor_type<storage_buffer>, operand_segment_sizes = array<i32: 1, 0>, set = 0 : index} : (index) -> !flow.dispatch.tensor<readonly:tensor<1x2x1x5xf32>>
      %8 = "hal.interface.binding.subspan"(%0) {alignment = 64 : index, binding = 1 : index, descriptor_type = #hal.descriptor_type<storage_buffer>, operand_segment_sizes = array<i32: 1, 0>, set = 0 : index} : (index) -> !flow.dispatch.tensor<writeonly:tensor<2x3x5xf32>>
      %9 = "hal.interface.workgroup.id"() {dimension = 0 : index} : () -> index
      %10 = "hal.interface.workgroup.count"() {dimension = 0 : index} : () -> index
      %11 = "hal.interface.workgroup.id"() {dimension = 1 : index} : () -> index
      %12 = "hal.interface.workgroup.count"() {dimension = 1 : index} : () -> index
      %13 = "hal.interface.workgroup.id"() {dimension = 2 : index} : () -> index
      %14 = "hal.interface.workgroup.count"() {dimension = 2 : index} : () -> index
      "scf.for"(%13, %2, %14) ({
      ^bb0(%arg0: index):
        "scf.for"(%11, %1, %12) ({
        ^bb0(%arg1: index):
          "scf.for"(%9, %3, %10) ({
          ^bb0(%arg2: index):
            %15 = "flow.dispatch.tensor.load"(%7, %arg0, %arg1, %arg2, %1) {operand_segment_sizes = array<i32: 1, 0, 3, 1, 0>, static_offsets = array<i64: 0, -9223372036854775808, -9223372036854775808, -9223372036854775808>, static_sizes = array<i64: -9223372036854775808, 1, 1, 1>, static_strides = array<i64: 1, 1, 1, 1>} : (!flow.dispatch.tensor<readonly:tensor<1x2x1x5xf32>>, index, index, index, index) -> tensor<?x1x1x1xf32>
            %16 = "tensor.empty"() : () -> tensor<1x1x1xf32>
            %17 = "linalg.fill"(%6, %16) ({
            ^bb0(%arg3: f32, %arg4: f32):
              "linalg.yield"(%arg3) : (f32) -> ()
            }) {operand_segment_sizes = array<i32: 1, 1>} : (f32, tensor<1x1x1xf32>) -> tensor<1x1x1xf32>
            %18 = "iree_linalg_ext.scatter"(%15, %5, %17) ({
            ^bb0(%arg3: f32, %arg4: f32):
              %19 = "arith.addf"(%arg4, %arg3) {fastmath = #arith.fastmath<none>} : (f32, f32) -> f32
              "iree_linalg_ext.yield"(%19) : (f32) -> ()
            }) {dimension_map = array<i64: 1>, lowering_config = #iree_codegen.lowering_config<tile_sizes = [[0, 1, 1, 1]]>, operand_segment_sizes = array<i32: 2, 1>, unique_indices = true} : (tensor<?x1x1x1xf32>, tensor<1x1xi32>, tensor<1x1x1xf32>) -> tensor<1x1x1xf32>
            "flow.dispatch.tensor.store"(%18, %8, %arg0, %arg1, %arg2) {operand_segment_sizes = array<i32: 1, 1, 0, 3, 0, 0>, static_offsets = array<i64: -9223372036854775808, -9223372036854775808, -9223372036854775808>, static_sizes = array<i64: 1, 1, 1>, static_strides = array<i64: 1, 1, 1>} : (tensor<1x1x1xf32>, !flow.dispatch.tensor<writeonly:tensor<2x3x5xf32>>, index, index, index) -> ()
            "scf.yield"() : () -> ()
          }) : (index, index, index) -> ()
          "scf.yield"() : () -> ()
        }) : (index, index, index) -> ()
        "scf.yield"() : () -> ()
      }) : (index, index, index) -> ()
      "func.return"() : () -> ()
    }) {function_type = () -> (), sym_name = "_main_dispatch_2_fill_2x3x5"} : () -> ()
  }) : () -> ()
  "hal.executable.variant_end"() : () -> ()
}) {sym_name = "embedded_elf_x86_64", target = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>} : () -> ()
/tmp/asr_toy_grad.mlir:16:10: error: failed to serialize executables
    %9 = "stablehlo.scatter"(%8, %1, %6) ({
         ^
/tmp/asr_toy_grad.mlir:16:10: note: see current operation:
"hal.executable"() ({
  "hal.executable.variant"() ({
    "hal.executable.export"() ({
    ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index, %arg3: index, %arg4: index):
      "hal.return"(%arg4, %arg3, %arg2) : (index, index, index) -> ()
    }) {layout = #hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer>]>]>, ordinal = 0 : index, sym_name = "_main_dispatch_2_fill_2x3x5", translation_info = #iree_codegen.translation_info<CPUDefault>} : () -> ()
    "builtin.module"() ({
      "func.func"() ({
        %0 = "arith.constant"() {value = 0 : index} : () -> index
        %1 = "arith.constant"() {value = 1 : index} : () -> index
        %2 = "arith.constant"() {value = 2 : index} : () -> index
        %3 = "arith.constant"() {value = 5 : index} : () -> index
        %4 = "arith.constant"() {value = 128 : index} : () -> index
        %5 = "arith.constant"() {value = dense<0> : tensor<1x1xi32>} : () -> tensor<1x1xi32>
        %6 = "arith.constant"() {value = 0.000000e+00 : f32} : () -> f32
        %7 = "hal.interface.binding.subspan"(%4) {alignment = 64 : index, binding = 0 : index, descriptor_type = #hal.descriptor_type<storage_buffer>, operand_segment_sizes = array<i32: 1, 0>, set = 0 : index} : (index) -> !flow.dispatch.tensor<readonly:tensor<1x2x1x5xf32>>
        %8 = "hal.interface.binding.subspan"(%0) {alignment = 64 : index, binding = 1 : index, descriptor_type = #hal.descriptor_type<storage_buffer>, operand_segment_sizes = array<i32: 1, 0>, set = 0 : index} : (index) -> !flow.dispatch.tensor<writeonly:tensor<2x3x5xf32>>
        %9 = "hal.interface.workgroup.id"() {dimension = 0 : index} : () -> index
        %10 = "hal.interface.workgroup.count"() {dimension = 0 : index} : () -> index
        %11 = "hal.interface.workgroup.id"() {dimension = 1 : index} : () -> index
        %12 = "hal.interface.workgroup.count"() {dimension = 1 : index} : () -> index
        %13 = "hal.interface.workgroup.id"() {dimension = 2 : index} : () -> index
        %14 = "hal.interface.workgroup.count"() {dimension = 2 : index} : () -> index
        "scf.for"(%13, %2, %14) ({
        ^bb0(%arg0: index):
          "scf.for"(%11, %1, %12) ({
          ^bb0(%arg1: index):
            "scf.for"(%9, %3, %10) ({
            ^bb0(%arg2: index):
              %15 = "flow.dispatch.tensor.load"(%7, %arg0, %arg1, %arg2, %1) {operand_segment_sizes = array<i32: 1, 0, 3, 1, 0>, static_offsets = array<i64: 0, -9223372036854775808, -9223372036854775808, -9223372036854775808>, static_sizes = array<i64: -9223372036854775808, 1, 1, 1>, static_strides = array<i64: 1, 1, 1, 1>} : (!flow.dispatch.tensor<readonly:tensor<1x2x1x5xf32>>, index, index, index, index) -> tensor<?x1x1x1xf32>
              %16 = "tensor.empty"() : () -> tensor<1x1x1xf32>
              %17 = "linalg.fill"(%6, %16) ({
              ^bb0(%arg3: f32, %arg4: f32):
                "linalg.yield"(%arg3) : (f32) -> ()
              }) {operand_segment_sizes = array<i32: 1, 1>} : (f32, tensor<1x1x1xf32>) -> tensor<1x1x1xf32>
              %18 = "iree_linalg_ext.scatter"(%15, %5, %17) ({
              ^bb0(%arg3: f32, %arg4: f32):
                %19 = "arith.addf"(%arg4, %arg3) {fastmath = #arith.fastmath<none>} : (f32, f32) -> f32
                "iree_linalg_ext.yield"(%19) : (f32) -> ()
              }) {dimension_map = array<i64: 1>, lowering_config = #iree_codegen.lowering_config<tile_sizes = [[0, 1, 1, 1]]>, operand_segment_sizes = array<i32: 2, 1>, unique_indices = true} : (tensor<?x1x1x1xf32>, tensor<1x1xi32>, tensor<1x1x1xf32>) -> tensor<1x1x1xf32>
              "flow.dispatch.tensor.store"(%18, %8, %arg0, %arg1, %arg2) {operand_segment_sizes = array<i32: 1, 1, 0, 3, 0, 0>, static_offsets = array<i64: -9223372036854775808, -9223372036854775808, -9223372036854775808>, static_sizes = array<i64: 1, 1, 1>, static_strides = array<i64: 1, 1, 1>} : (tensor<1x1x1xf32>, !flow.dispatch.tensor<writeonly:tensor<2x3x5xf32>>, index, index, index) -> ()
              "scf.yield"() : () -> ()
            }) : (index, index, index) -> ()
            "scf.yield"() : () -> ()
          }) : (index, index, index) -> ()
          "scf.yield"() : () -> ()
        }) : (index, index, index) -> ()
        "func.return"() : () -> ()
      }) {function_type = () -> (), sym_name = "_main_dispatch_2_fill_2x3x5"} : () -> ()
    }) : () -> ()
    "hal.executable.variant_end"() : () -> ()
  }) {sym_name = "embedded_elf_x86_64", target = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>} : () -> ()
  "hal.executable_end"() : () -> ()
}) {sym_name = "_main_dispatch_2", sym_visibility = "private"} : () -> ()
```

### What component(s) does this issue relate to?

Compiler

### Version information

fca4102f97ba7e9795ff0dd1489653e2214867d6

**Comment by rsuderman** [comment]
Looks like the issue is quite minor. We do some cleanup that allows us to infer shapes on most `tensor.extract` operations. One does not infer a dynamic shape is static so it leaves the shape as dynamic causing a mismatch. Test passes if we ignore the verifier error so I am going to fix up the tiling code to infer shapes more correctly.
